### DISCUSSION TITLE
Generate .yaml file for VSCode extention "Shell Script Command Completion"

### DISCUSSION THREAD

**Discussion by vdesabou** [bashly-framework/bashly#437] [Ideas]
Hello!

I'm using a lot the CLI generated by bashly in shell scripts which I edit with VSCode.

There is this extension https://marketplace.visualstudio.com/items?itemName=tetradresearch.vscode-h2o that provides intellisense completion.

It seems to work by reading .yaml files with following [format](https://github.com/yamaton/h2o-curated-data#data-structure). 

Here is an example for [docker](https://github.com/yamaton/h2o-curated-data/blob/main/general/yaml/docker.yaml).

I was wondering if there would be an easy way to generate such yaml file using bashly ?

**Comment by DannyBen** [comment]
There could be a way to do this, and possibly quite easily.

Did you see the [Rendering Documentation](https://bashly.dannyb.co/advanced/rendering/) section?
This feature was designed to let you generate documentation (markdown, man pages, etc.), but it can be used to render anything.

You can use it to render your YAML files.

The workflow should be simple:
1. You create a folder with a template in your project dir.
2. The template has access to the data structure of your script.
3. You then run `bashly render` to generate the output.

Read the documentation, let me know if it helps or if you need assistance with it.

**Reply by vdesabou** [reply]
Thanks for the pointer @DannyBen !
I give it a try and that will do exactly what I need.

**Reply by DannyBen** [reply]
Sure thing. Feel free to ask for assistance if it's not going as easily as you expect.
If it is not intuitive, it's a bug. 😉

**Reply by vdesabou** [reply]
All good ! I managed to get it work [here](https://github.com/vdesabou/kafka-docker-playground/tree/master/scripts/cli/templates/shell-script-command-completion)
It's ugly code but [end result](https://github.com/vdesabou/kafka-docker-playground/blob/master/scripts/cli/playground.yaml) is what I needed to generate 

Thanks again for your awesome project and all the help you provide every time ç

**Reply by DannyBen** [reply]
Nive work.
You know that there is no requirement to use the gtx template. You can just use plain Ruby if you want, or any other templating language in Ruby.

In particular, since you are rendering a YAML file, you don't need any templating language.
You can simply create your data object (`data = { 'name' => 'playground', ...}`) and convert it to yaml with `data.to_yaml`.

**Comment by DannyBen** [comment]
Vincent - I took the liberty to prepare a simple "convert to YAML" render script. Try this instead of your `render.rb` - I think this approach will be much simpler for you to get a nice YAML.

I already noticed you have some [indentation problems](https://github.com/vdesabou/kafka-docker-playground/blob/a266b6442051680323194723beceb251ccb09809/scripts/cli/playground.yaml#L165-L166) in your output, which demonstrates why generating YAML with a template is error prone.

### `render.rb`

```ruby
# render script - yaml example
require 'yaml'
require 'lp'

# Convert command to our desired data structure
def convert_command(input_command)
  # Update this structure as needed
  result = {
    'name'        => input_command.name,
    'description' => input_command.help,
    'usage'       => input_command.usage_string
  }

  # Call this function recureively for its subcommands
  input_command.commands.each do |subcommand|
    result['subcommands'] ||= []
    result['subcommands'] << convert_command(subcommand)
  end

  result
end

# Prepare data structure
data = convert_command(command)

# Save the output
save "#{target}/#{command.name}.yaml", data.to_yaml

# Show the YAML if requested with --show
lp data if show

```

with this approach, you can get rid of any GTX template, and be sure you get a nice, valid YAML output.
